File: caesarCipher.pas
// Brian Dones
// Alan Labouseur
// Theory of Programming Languages
// 26 February 2014

// Programming in the Past (Pascal) in a procedural manner

program caesarCipher;

Var
  text : string;
  shiftValue : integer; // Shift Value given for the string
  i : integer; // this is used for all the procedures and 
  helper : integer; // this is needed for the solve function

  
// Since there is the possibility of having Capitalizecase letters and lowercase letters in a string
// I need to take the string in one function and return a new string that is either all
// capitalized or lowercased. That's the purpose of my NewString function. By subtracting 32 from
// the ascii value, I can turn all the lowercase letter values into their equivalent Capitalizecase
// value.


function Capitalize (text : string) : string;
begin
     for i := 1 to Length(text) do begin
       if (ord(text[i]) > 90) then
         text[i] := chr(ord(text[i]) - 32)
     end;
     Capitalize := text;
end;

procedure Encrypt (text : string; shiftValue : integer);
begin
     text := Capitalize(text);
     for i := 1 to Length(text) do begin
       if (ord(text[i]) = 32) then
         text[i] := text[i]
       else
           if ((ord(text[i]) + shiftValue) > 90) then begin
             text[i] := chr(64 + (ord(text[i]) + shiftValue - 90));
           end
           else
             text[i] := chr(ord(text[i]) + shiftValue);
     end;
     writeln(text);
end;

procedure Decrypt (text : string; shiftValue : integer);
begin
     text := Capitalize(text);
     for i := 1 to Length(text) do begin
       if (ord(text[i]) = 32) then
         text[i] := text[i]
       else
           if ((ord(text[i]) - shiftValue) < 65) then begin
              text[i] := chr(91 - (65 - (ord(text[i]) - shiftValue)))
           end
           else
             text[i] := chr(ord(text[i]) - shiftValue);
       end;
     writeln(text)
end;

procedure Solve (text : string);
begin
     for helper := 26 downto 0 do begin
       Decrypt(text, helper);
     end
end;

begin
     writeln('Encrypt: "Sean Connery is the best James Bond", 8');
     Encrypt('Sean Connery is the best James Bond', 8);
     writeln('Decrypt: "This is a test case", 5');
     Decrypt('This is a test case', 5);
     writeln('Solve: "This is a final test case", 26');
     Solve('This is a final test case');
end.

File: caesarCipher.f
program caesarCipher
	implicit none
	
	interface
		subroutine encrypt (text, shiftValue)
			character (*), intent(in) :: text
			integer :: shiftValue
		end subroutine encrypt
		
		
		subroutine decrypt (text, shiftValue)
			character (*), intent(in) :: text
			integer :: shiftValue
		end subroutine decrypt
		
		
		subroutine solve (text, shiftValue)
			character (*), intent(in) :: text
			integer :: shiftValue
		end subroutine solve
		
		
		function capitalize (text)
			character (*), intent(in) :: text
			character (len=len_trim(text)) :: capitalize
		end function capitalize
		
	end interface
	
	character (len = 35) :: firstTestCase ="Sean Connery is the best James Bond"
	character (len = 19) :: secondTestCase ="This is a test case"
	character (len = 25) :: thirdTestCase ="This is a final test case"
	integer :: shiftValue
	shiftValue = 8

	firstTestCase = capitalize(firstTestCase)
	secondTestCase = capitalize(secondTestCase)
	thirdTestCase = capitalize(thirdTestCase)

	print *, "String: Sean Connery is the best James Bond, Shift: 8"
	call encrypt(firstTestCase, shiftValue)

	print *, ""
	print *, "String: This is a test case, Shift Value: 19"
	shiftValue = 19
	call decrypt(secondTestCase, shiftValue)

	print *, ""
	print *, "String: This is a final test case, Shift Value: 26"
	shiftValue = 26
	call solve(thirdTestCase, shiftValue)


end program caesarCipher

function capitalize(text) result(tempValue)
	implicit none
	character(len=*), intent(in) :: text
	character(len=len_trim(text)):: tempValue
	integer :: i
	tempValue = ""


	do i = 1, len(text), 1
		if (ichar(text(i:i)) > 90) then
			tempValue(i:i) = achar(iachar(text(i:i)) - 32)
		else
			tempValue(i:i) = text(i:i)
		endif
	end do
end function capitalize


subroutine encrypt(text, shiftValue)
	implicit none
	character (len=*), intent(in) :: text
	character (len=len(text)) :: tempValue
	integer :: shiftValue
	integer :: i

	tempValue = ""


	do i = 1, len(text), 1
		if (IACHAR(text(i:i)) == 32) then
			tempValue(i:i) = " "
		else 
			if (IACHAR(text(i:i)) + shiftValue > 90 ) then
				tempValue(i:i) = ACHAR(64 + ((IACHAR(text(i:i)) + shiftValue) - 90))
			else
				tempValue(i:i) = ACHAR(IACHAR(text(i:i)) + shiftValue)
			endif
		endif
	end do

	print *, tempValue
end subroutine encrypt

subroutine decrypt(text, shiftValue)
	implicit none
	character (len=*), intent(in) :: text
	character (len=len(text)) :: tempValue
	integer :: shiftValue
	integer :: i

	tempValue = ""


	do i = 1, len(text), 1
		if (IACHAR(text(i:i)) == 32) then
			tempValue(i:i) = " "
		else 
			if (IACHAR(text(i:i)) - shiftValue < 65 ) then
				tempValue(i:i) = ACHAR(91 - (65 - (IACHAR(text(i:i)) - shiftValue)))
			else
				tempValue(i:i) = ACHAR(IACHAR(text(i:i)) - shiftValue)
			endif
		endif
	end do

	print *, tempValue
end subroutine decrypt

subroutine solve(text, shiftValue)
	implicit none
	character (len=*), intent(in) :: text
	integer :: shiftValue
	integer :: i

	do i = shiftValue, 0, -1 
		call decrypt(text, i)
	end do
end subroutine solve